File: ported/auto/40ab58.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/40ab58.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { DestroyWindow, _memset } from "../../runtime/win32.js";
import { callIndirect } from "../../runtime/win32/context.js";
import { regs } from "../../runtime/regs.js";
import { FUN_00413470 } from "./413470.js";
export function FUN_0040ab58(heap) {
  const __sp = heap.allocFrame(112);
  const __addr_local_74 = __sp + 0;
  const __addr_local_70 = __sp + 4;
  const __addr_local_c = __sp + 104;
  const __addr_local_8 = __sp + 108;
  try {
  let puVar1 = 0;
  let iVar2 = 0;
  heap.setU32(__addr_local_8, (heap.u32(0x005ebf48)) >>> 0);
  _memset(heap, __addr_local_74, 0, 0x6c);
  heap.setU32(__addr_local_74, (0x6c) >>> 0);
  heap.setU32(__addr_local_70, (1) >>> 0);
  iVar2 = (((regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf34)) + 0x58)), heap.u32(0x005ebf34), __addr_local_74))) >>> 0);
  if ((iVar2 == 0) && ((heap.u32(__addr_local_c) & 0x10) != 0)) {
    (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf30)) + 0x28)), heap.u32(0x005ebf30)));
  }
  while (heap.u32(__addr_local_8) != 0x0) {
    puVar1 = ((heap.u32(heap.u32(__addr_local_8) + (2) * 4)) >>> 0);
    (regs.eax = FUN_00413470(heap, heap.u32(heap.u32(__addr_local_8))));
    (regs.eax = callIndirect(heap, heap.u32((heap.i32(heap.u32(heap.u32(__addr_local_8) + (1) * 4)) + 8)), heap.u32(heap.u32(__addr_local_8) + (1) * 4)));
    (regs.eax = FUN_00413470(heap, heap.u32(__addr_local_8)));
    heap.setU32(__addr_local_8, (puVar1) >>> 0);
  }
  heap.setU32(0x005ebf48, (0x0) >>> 0);
  if (heap.u32(0x005ebf44) != 0x0) {
    (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf44)) + 8)), heap.u32(0x005ebf44)));
    heap.setU32(0x005ebf44, (0x0) >>> 0);
  }
  if (heap.u32(0x005ebf38) != 0) {
    (regs.eax = FUN_00413470(heap, heap.u32(0x005ebf38)));
    heap.setU32(0x005ebf38, (0) >>> 0);
  }
  if (heap.u32(0x005ebf34) != 0x0) {
    (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf34)) + 8)), heap.u32(0x005ebf34)));
    heap.setU32(0x005ebf34, (0x0) >>> 0);
  }
  if (heap.u32(0x005ebf3c) != 0x0) {
    (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf3c)) + 8)), heap.u32(0x005ebf3c)));
    heap.setU32(0x005ebf3c, (0x0) >>> 0);
  }
  (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf30)) + 0x4c)), heap.u32(0x005ebf30)));
  (regs.eax = callIndirect(heap, heap.u32((heap.u32(heap.u32(0x005ebf30)) + 0x50)), heap.u32(0x005ebf30), heap.u32(0x005e916c), 8));
  if (heap.u32(0x005e916c) != ((0x0) | 0)) {
    DestroyWindow(heap, heap.u32(0x005e916c));
    heap.setU32(0x005e916c, (((0x0) | 0)) >>> 0);
  }
  return;
} finally {
    heap.freeFrame(112);
  }
}

File: decompiled/c/40ab58.c

void FUN_0040ab58(void)

{
  undefined4 *puVar1;
  int iVar2;
  undefined4 local_74;
  undefined4 local_70;
  byte local_c;
  undefined4 *local_8;
  
  local_8 = DAT_005ebf48;
  _memset(&local_74,0,0x6c);
  local_74 = 0x6c;
  local_70 = 1;
  iVar2 = (**(code **)(*DAT_005ebf34 + 0x58))(DAT_005ebf34,&local_74);
  if ((iVar2 == 0) && ((local_c & 0x10) != 0)) {
    (**(code **)(*DAT_005ebf30 + 0x28))(DAT_005ebf30);
  }
  while (local_8 != (undefined4 *)0x0) {
    puVar1 = (undefined4 *)local_8[2];
    FUN_00413470(*local_8);
    (**(code **)(*(int *)local_8[1] + 8))(local_8[1]);
    FUN_00413470(local_8);
    local_8 = puVar1;
  }
  DAT_005ebf48 = (undefined4 *)0x0;
  if (DAT_005ebf44 != (int *)0x0) {
    (**(code **)(*DAT_005ebf44 + 8))(DAT_005ebf44);
    DAT_005ebf44 = (int *)0x0;
  }
  if (DAT_005ebf38 != 0) {
    FUN_00413470(DAT_005ebf38);
    DAT_005ebf38 = 0;
  }
  if (DAT_005ebf34 != (int *)0x0) {
    (**(code **)(*DAT_005ebf34 + 8))(DAT_005ebf34);
    DAT_005ebf34 = (int *)0x0;
  }
  if (DAT_005ebf3c != (int *)0x0) {
    (**(code **)(*DAT_005ebf3c + 8))(DAT_005ebf3c);
    DAT_005ebf3c = (int *)0x0;
  }
  (**(code **)(*DAT_005ebf30 + 0x4c))(DAT_005ebf30);
  (**(code **)(*DAT_005ebf30 + 0x50))(DAT_005ebf30,DAT_005e916c,8);
  if (DAT_005e916c != (HWND)0x0) {
    DestroyWindow(DAT_005e916c);
    DAT_005e916c = (HWND)0x0;
  }
  return;
}

